src/Compilers/Core/AnalyzerDriver/AnalyzerExecutor.cs

```diff
@@ -44,8 +44,10 @@ internal class AnalyzerExecutor
         /// <param name="isCompilerAnalyzer">Delegate to determine if the given analyzer is compiler analyzer. 
         /// We need to special case the compiler analyzer at few places for performance reasons.</param>
         /// <param name="analyzerManager">Analyzer manager to fetch supported diagnostics.</param>
-        /// <param name="singleThreadedAnalyzerToGateMap">Map from non-thread safe analyzers to unique gate objects. 
-        /// All analyzer callbacks for these analyzers will be guarded with a lock on the gate.
+        /// <param name="getAnalyzerGate">
+        /// Delegate to fetch the gate object to guard all callbacks into the analyzer.
+        /// It should return a unique gate object for the given analyzer instance for non-concurrent analyzers, and null otherwise.
```

Comment:
Even concurrent analyzers will have some locking requirements, to prevent executing related actions concurrently. It might be prudent to always provide a gate object and just not use it for actions with no relatives, or we might need a rather different mechanism. (Actions unrelated to each other but each related to other actions can't be optimally managed with a single gate, and might need one gate per relationship.)

If a concurrent analyzer has no related actions, then it requires no gating and the comment and implied mechanism are fine as written.

I guess I'm just claiming that more thought is necessary here, and mentioning concurrent analyzers might be misleading.